File: lua/samples.lua
-- Consts and prep
PI = 3.14159265358979323846;
NORMAL95CONFIDENCE = 1.6448536269514722;
math.randomseed(1234)

-- Random distribution functions
function sample_normal_0_1()
  local u1 = math.random()
	local u2 = math.random()
	local result = math.sqrt(-2 * math.log(u1)) * math.sin(2 * PI * u2)
	return result
end

function sample_normal(mean, sigma)
	return mean + (sigma * sample_normal_0_1())
end

function sample_uniform(min, max)
  return math.random() * (max - min) + min
end

function sample_lognormal(logmean, logsigma)
	return math.exp(sample_normal(logmean, logsigma))
end

function sample_to(low, high)
	local loglow = math.log(low);
	local loghigh = math.log(high);
	local logmean = (loglow + loghigh) / 2;
	local logsigma = (loghigh - loglow) / (2.0 * NORMAL95CONFIDENCE);
	return sample_lognormal(logmean, logsigma, seed);
end

-- Mixture
function mixture(samplers, weights, n)
	assert(#samplers == #weights)
  local l = #weights
	local sum_weights = 0
	for i = 1, l, 1 do
		-- ^ arrays start at 1
		sum_weights = sum_weights + weights[i]
	end
  local cumsummed_normalized_weights = {}
	table.insert(cumsummed_normalized_weights, weights[1]/sum_weights)
	for i = 2, l, 1 do
		table.insert(cumsummed_normalized_weights, cumsummed_normalized_weights[i-1] + weights[i]/sum_weights)
	end
	
	local result = {}
	for i = 1, n, 1 do
		r = math.random()
		local i = 1
		while r > cumsummed_normalized_weights[i] do
			i = i+1
		end
		table.insert(result, samplers[i]())
	end
	return result
end


-- Main 
p_a = 0.8
p_b = 0.5
p_c = p_a * p_b

function sample_0() return 0 end
function sample_1() return 1 end
function sample_few() return sample_to(1, 3) end
function sample_many() return sample_to(2, 10) end

samplers = {sample_0, sample_1, sample_few, sample_many}
weights = { (1 - p_c), p_c/2, p_c/4, p_c/4 }

n = 1000000
result = mixture(samplers, weights, n)
sum = 0
for i = 1, n, 1 do
	sum = sum + result[i]
	-- print(result[i])
end
print(sum/n)

File: ocaml/samples.ml
(* Constants *)
let pi = acos (-1.)
let normal_95_ci_length = 1.6448536269514722

(* List manipulation helpers *)
let sumFloats xs = List.fold_left(fun acc x -> acc +. x) 0.0 xs

let normalizeXs xs = 
  let sum_xs = sumFloats xs in 
  List.map(fun x -> x /. sum_xs) xs

let cumsumXs xs = 
  let _, cum_sum = List.fold_left(fun (sum, ys) x -> 
    let new_sum = sum +. x in
    new_sum, ys @ [new_sum]
  ) (0.0, []) xs in
  cum_sum

let rec nth xs (n: int) = 
  match xs with
  | [] -> Error "nth function finds no match"
  | y :: ys -> if n = 0 then Ok(y) else nth ys (n-1)
  (* 
     Note that this is O(n) access.
     That is the cost of using the nice match syntax,
     which is not possible with OCaml arrays
  *)

let findIndex xs test = 
  let rec recursiveHelper ys i = 
    match ys with
      | [] -> Error "findIndex doesn't find an index"
      | z :: zs -> if test z then Ok(i) else recursiveHelper zs (i+1)
    in
  recursiveHelper xs 0

let unwind xs = 
  let rec tailRecursiveHelper ys acc =
    match ys with
    | [] -> Ok(acc)
    | Error e :: _ -> Error e
    | Ok(y) :: ys -> tailRecursiveHelper ys (y :: acc) 
  in 
  tailRecursiveHelper xs []
  
let unwindSum xs = 
  let rec tailRecursiveHelper ys sum =
    match ys with
    | [] -> Ok(sum)
    | Error e :: _ -> Error e
    | Ok(y) :: ys -> tailRecursiveHelper ys (y +. sum) 
  in 
  tailRecursiveHelper xs 0.0

(* Array helpers *)
let unwindSumArray xs = 
  Array.fold_left(fun acc x -> 
    (
      match acc, x with
      | Error e, _ -> Error e
      | _, Error e -> Error e
      | Ok(sum), Ok(y) -> Ok(sum +. y)
    )
  ) (Ok 0.0) xs

let sumFloats xs = List.fold_left(fun acc x -> acc +. x) 0.0 xs

(* Basic samplers *)
let sampleZeroToOne () : float = Random.float 1.0

let sampleStandardNormal (): float = 
  let u1 = sampleZeroToOne () in 
  let u2 = sampleZeroToOne () in 
  let z = sqrt(-2.0 *. log(u1)) *. sin(2.0 *. pi *. u2) in 
  z

let sampleNormal mean std = mean +. std *. (sampleStandardNormal ())

let sampleLognormal logmean logstd = exp(sampleNormal logmean logstd)

let sampleTo low high = 
  let loglow = log(low) in
  let loghigh = log(high) in
  let logmean = (loglow +. loghigh) /. 2.0 in
  let logstd = (loghigh -. loglow) /. (2.0 *. normal_95_ci_length ) in
  sampleLognormal logmean logstd

let mixture (samplers: (unit -> float) list) (weights: float list): (float, string) result = 
   if (List.length samplers <> List.length weights) 
    then Error "in mixture function, List.length samplers != List.length weights"
  else
    let normalized_weights = normalizeXs weights in
    let cumsummed_normalized_weights = cumsumXs normalized_weights in
    let p = sampleZeroToOne () in
    let chosenSamplerIndex = findIndex cumsummed_normalized_weights (fun x -> p < x) in
    let sampler = match chosenSamplerIndex with
      | Error e -> Error e
      | Ok(i) -> nth samplers i
    in
    let sample = match sampler with
      | Error e -> Error e
      | Ok(f) -> Ok(f ())
    in 
    sample

let () =
  let sample0 () = 0. in 
  let sample1 () = 1. in
  let sampleFew () = sampleTo 1. 3. in
  let sampleMany () = sampleTo 2. 10. in 
  let p1 = 0.8 in 
  let p2 = 0.5 in 
  let p3 = p1 *. p2 in 
  let weights = [ 1. -. p3; p3 /. 2.; p3 /. 4.; p3/. 4. ] in
  let sampler () = mixture [ sample0; sample1; sampleFew; sampleMany ] weights in
  let n = 1_000_000 in 
  let samples = Array.init n (fun _ -> sampler ()) in 
  match unwindSumArray samples with
    | Error err -> Printf.printf "Error %s\n" err
    | Ok(sum) -> (
        let mean = sum /. float_of_int(n) in 
        Printf.printf "Mean: %f\n" mean
      )
